File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    public class Word {
        private final String name;
        private final int cnt;

        public Word(String name, int cnt) {
            this.name = name;
            this.cnt = cnt;
        }
    }
    
    public int solution(String begin, String target, String[] words) {
        Queue<Word> que = new LinkedList<>();
        boolean[] visited = new boolean[words.length];

        que.offer(new Word(begin, 0));

        while (!que.isEmpty()) {
            Word word = que.poll();

            for (int i = 0; i < words.length; i++) {
                if (!visited[i]) {
                    int cnt = 0;

                    for (int j = 0; j < target.length(); j++) {
                        if (word.name.charAt(j) != words[i].charAt(j)) {
                            cnt++;
                        }
                    }

                    if (cnt == 1) {
                        if (words[i].equals(target)) {
                            return word.cnt + 1;
                        }

                        que.offer(new Word(words[i], word.cnt + 1));
                        visited[i] = true;
                    }
                }
            }
        }

        return 0;
    }
}

File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.kt
import java.util.*

class Solution {
    fun solution(begin: String, target: String, words: Array<String>): Int {
        val que: Queue<Pair<String, Int>> = LinkedList()
        val visited = BooleanArray(words.size)
        var answer = 0

        que.add(Pair(begin, 0))

        loop@ while (que.isNotEmpty()) {
            val (now, cnt) = que.poll()

            for (i in words.indices) {
                var temp = 0

                if (visited[i].not()) {
                    for (j in begin.indices) {
                        if (now[j] != words[i][j]) {
                            temp++
                        }
                    }
                }

                if (temp == 1) {
                    if (words[i] == target) {
                        answer = cnt + 1
                        break@loop
                    }

                    que.add(Pair(words[i], cnt + 1))
                    visited[i] = true
                }
            }
        }

        return answer
    }
}